ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,250 @@
+                                       .EnsureInlayHintsDefault(settingsStore) }
+
+    member x.IsDisabled =
+        x.TopLevelMembers = PushToHintMode.Never &&
+        x.LocalBindings = PushToHintMode.Never
+
+
+type private MembersVisitor(settings) =
+    inherit TreeNodeVisitor<NodesRequiringHints>()
+
+    let isTopLevelMember (node: ITreeNode) =
+        match node with
+        | :? ITopBinding
+        | :? IMemberDeclaration
+        | :? IConstructorDeclaration -> true
+        | _ -> false
+
+    override x.VisitNode(node, context) =
+        if settings.LocalBindings = PushToHintMode.Never && isTopLevelMember node then () else
+
+        for child in node.Children() do
+            if settings.TopLevelMembers = PushToHintMode.Never &&
+               isTopLevelMember child then x.VisitNode(child, context) else
+
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitLambdaExpr(lambda, context) =
+        let result = collectTypeHintAnchorsForLambda lambda
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(lambda, context)
+
+    override x.VisitLocalBinding(binding, context) =
+        let result = collectTypeHintAnchorsForBinding binding
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitTopBinding(binding, context) =
+        let result = collectTypeHintAnchorsForBinding binding
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitMemberDeclaration(memberDecl, context) =
+        let result = collectTypeHintsAnchorsForMember memberDecl
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(memberDecl, context)
+
+    override x.VisitPrimaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitSecondaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitForEachExpr(forEachExpr, context) =
+        let result = collectTypeHintAnchorsForEachExpr forEachExpr
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(forEachExpr, context)
+
+type private PatternsHighlightingProcess(logger: ILogger, fsFile, settingsStore: IContextBoundSettingsStore,
+                                           daemonProcess: IDaemonProcess, settings) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let createTypeHintHighlighting (fcsType: FSharpType) (displayContext: FSharpDisplayContext) range pushToHintMode =
+        let displayContext = displayContext.WithShortTypeNames(true)
```

Comment:
I wonder if we could reuse the display context between symbols. We use short names anyway, so the context is not needed (much?), as the main thing it brings here is what namespaces are open.

Reply:
Maybe we could just use a default context with short names?

Reply:
If there's one, and it works with it, then, for sure, let's do it.